File: backend/dataset/rotate/rotate_1.py
class Solution:    
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        length = len(nums)
        k = k % length
        return self.withConstantSpaceAndOptimized(nums, k)
    
    # Time complexity: O(n), Space complexity: O(1)
    def withConstantSpaceAndOptimized(self, nums, k):
            for index in range(0, k + 1):
                pos = (index + k) % len(nums)
                nums[pos], nums[index] = nums[index], nums[pos]
            return nums
 

File: backend/CDataset/rotate/rotate_1.cpp
class Solution {
public:
    #include<vector>
    void rotate(vector<int>& nums, int k) {
        reverse(nums.begin(),nums.end());                   // reverse the array
        reverse(nums.begin(),nums.begin()+k%nums.size());  // reverse the first k ele
        reverse(nums.begin()+k%nums.size(),nums.end());   // reverse the last n-k ele
    }
};
